layer/filestore.go

```diff
@@ -191,6 +201,23 @@ func (fms *fileMetadataStore) GetCacheID(layer ChainID) (string, error) {
 	return content, nil
 }
 
+func (fms *fileMetadataStore) GetForeignSource(layer ChainID) (*distribution.Descriptor, error) {
+	content, err := ioutil.ReadFile(fms.getLayerFilename(layer, "foreign-source"))
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var ref *distribution.Descriptor
+	err = json.Unmarshal(content, &ref)
+	if err != nil {
+		return nil, err
+	}
+	return ref, err
```

Comment:
If the content of `foreign-source` somehow got set to `null`, this could return a nil pointer. I think it would be a little safer to have `var ref distribution.Descriptor` and unmarshal into that. Then this function could return `distribution.Descriptor` instead of `*distribution.Descriptor` so there's not even any ambiguity about whether the result can be nil.